src/Symfony/Component/Messenger/Failure/SendFailedMessageToFailureTransportListener.php

```diff
@@ -48,35 +43,26 @@ public function onMessageFailed(WorkerMessageFailedEvent $event)
             return;
         }
 
-        $envelope = $event->getEnvelope();
-
-        // avoid re-sending to the failed sender
-        foreach ($envelope->all(SentStamp::class) as $sentStamp) {
-            /** @var SentStamp $sentStamp */
-            if ($sentStamp->getSenderAlias() === $this->failureSenderAlias) {
-                return;
-            }
-        }
-
-        // remove the received stamp so it's redelivered
         $throwable = $event->getThrowable();
         if ($throwable instanceof HandlerFailedException) {
             $throwable = $throwable->getNestedExceptions()[0];
         }
 
         $flattenedException = \class_exists(FlattenException::class) ? FlattenException::createFromThrowable($throwable) : null;
-        $envelope = $envelope->withoutAll(ReceivedStamp::class)
-            ->withoutAll(TransportMessageIdStamp::class)
-            ->with(new SentToFailureTransportStamp($throwable->getMessage(), $event->getReceiverName(), $flattenedException))
-            ->with(new RedeliveryStamp(0, $this->failureSenderAlias));
+        $failedMessage = new FailedMessage($event->getEnvelope(), $throwable->getMessage(), $flattenedException);
 
         if (null !== $this->logger) {
-            $this->logger->info('Rejected message {class} will be sent to the failure transport {transport}.', [
-                'class' => \get_class($envelope->getMessage()),
-                'transport' => $this->failureSenderAlias,
+            $this->logger->info('Rejected message {class} will be sent to the failure transport.', [
+                'class' => \get_class($event->getEnvelope()->getMessage()),
             ]);
         }
 
+        $envelope = new Envelope($failedMessage);
+        // route this new message to the same bus as the original message
+        if (null !== $busName = $this->getBusName($event->getEnvelope())) {
+            $envelope = $envelope->with(new BusNameStamp($busName));
```

Comment:
the null handling is useless currently. does it need to set the new stamp at all considering it usually sets the same BusNameStamp that it already had?

Reply:
the fallback to MessageBusInterface is already implemented in RoutableMessageBus. so this part really seems irrelevant

Reply:
Yea, I had a mistake in `getBusName()` that you caught.

The idea is:

If the original message has a `BusNameStamp`, add a `BusNameStamp` here so that the `FailedMessage` is dispatched through the same bus... which seems to make sense. If the original message doesn't have this stamp for some reason, do nothing, and allow `RoutableMessageBus` to handle the fallback to the "default" bus.

Make sense?

Reply:
I see. One is the failed message envelope and the other is the original envelope. Maybe renaming the vars to make it more obivious would be good.